src/v/datalake/translation/scheduling.cc

```diff
@@ -428,9 +434,12 @@ ss::future<> scheduler::main() {
           [this] { return requires_scheduling_actions(); });
         vlog(
           datalake_log.trace,
-          "scheduler tick,  memory_exhausted: {}",
-          _mem_tracker->memory_exhausted());
-        if (_mem_tracker->memory_exhausted()) {
+          "scheduler tick,  memory_exhausted: {} finish: {}",
+          _mem_tracker->memory_exhausted(),
+          _executor.translators_for_immediate_finish.size());
+        if (
+          _mem_tracker->memory_exhausted()
+          || !_executor.translators_for_immediate_finish.empty()) {
             co_await _scheduling_policy->on_resource_exhaustion(
```

Comment:
I'm squinting a bit trying to convince myself this call to on_resource_exhaustion never blocks indefinitely, waiting for a schedule_one_translation() call that isn't coming... Any reason for concern there?

Reply:
yeh, it's tightly coupled which isn't great. but it looks safe to me. we set the `finish bit` once, and then the translator will always reset it after a few seconds in its idle polling loop (see the `ss::defer` that clears it), or, if it is running/waiting it will receive the stop signal which brings it back to the same loop containing the same RAII device which clears the bit.

Reply:
from slackadoo

andrew

> like if all trnaslators are idle but take up space, we get to on_resource_exhaustion and set that the translator should finish, but the scheduler isn’t given the opportunity to kick it? (assuming schedule_one_translation is what finishes it, i forget)

noah

> idle translators are not scheduled into the running (or waiting and then running) state to finish because they are blocked polling for new offsets to translate. instead, that same polling loop notices the request to finish via set_finish_translation and skip all of the offset querying of the underlying partition, building a reader, etc... and goes directly to jail